File: Task4/src/Apostrophe.scala
/**
 * Created by Шолохов on 02.02.2015.
 */
case class Apostrophe (expression : Expression) extends Expression{
  override def evaluate(values: Map[String, Boolean]): Boolean = expression.evaluate(values)
  override def toString(): String = expression.toString() + "'"
  override def equals(o: Any): Boolean = {
    o match {
      case a : Apostrophe => a.expression.equals(expression)
      case _ => false
    }
  }
}

File: Task4/src/Apostrophe.java
public class Apostrophe extends Expression {
    public final Expression child;

    public Apostrophe(Expression child) {
        this.child = child;
    }

    public boolean equals(Object other) {
        if (other instanceof Apostrophe) {
            return child.equals(((Apostrophe) other).child);
        } else {
            return false;
        }
    }

    protected Expression substituteExpression(Expression where, Expression what) {
        return new Apostrophe(child.substituteTerm(where, what));
    }

    public String toString() {
        return "" + child + "'";
    }

    public int hashCode() {
        return toString().hashCode();
    }
}
